README.md
# KiteConnect

Elixir module for the Zerodha Kite Connect API

## Installation

Add `kite_connect` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:kite_connect, "~> 0.1"}
  ]
end
```

## Setup

Your project should be an OTP application. In your `application.ex`
file start `KiteConnect.State` as a worker.

```elixir
 # Define workers and child supervisors to be supervised
    children = [

      # Start your own worker
      worker(KiteConnect.State, []),
    ]

```

A sample iex session:

```
iex(1)> KiteConnect.init(your_api_key, your_api_secret)
:ok
iex(2)> KiteConnect.set_access_token(your_request_token)
:ok
iex(3)> KiteConnect.Quote.ltp("NSE:INFY")
724.3
```


CHANGELOG.md

# Changelog

## v0.1.7 Tue Feb 19 23:25:46 IST 2019

* Use KiteConnect.State to store API stuff

## v0.1.0 Wed Feb 13 17:49:54 IST 2019

* Release basic version



test/kite_connect_test.exs
defmodule KiteConnectTest do
  use ExUnit.Case

  setup do
    Application.put_env(:kite_connect_app, :kite_api_endpoint, 'https://api.kite.trade')
    Application.put_env(:kite_connect_app, :kite_api_key, 'xx')
    Application.put_env(:kite_connect_app, :kite_api_secret, 'yyy')
  end
  
  test "gen_url" do
    assert KiteConnect.gen_url("quote.ltp", "NSE:INFY") == "https://api.kite.trade/quote/ltp?i=NSE:INFY"
    assert KiteConnect.gen_url("session.token") == "https://api.kite.trade/session/token"

    assert KiteConnect.gen_url("instruments.historical", "NSE:INFY", "minute") ==
             "https://api.kite.trade/instruments/historical/NSE:INFY/minute"
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :kite_connect, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:kite_connect, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env()}.exs"


lib/kite_connect/quote.ex
defmodule KiteConnect.Quote do
  @moduledoc """
  Zerodha KiteConnect API for Market Quotes and Instruments
  """

  @doc """
  Get last traded price of instrument(s).
  Instruments can be provided as exchange:tradingsymbol combination
  or use instrument token
  
  KiteConnect.Quote.ltp("NSE:INFY")
  """
  def ltp(instruments) when is_list(instruments) do
    url = KiteConnect.gen_url("quote.ltp", instruments)
    headers = KiteConnect.gen_headers()
    res_body = KiteConnect.HTTP.get(url, headers)

    Enum.map(instruments, fn x -> %{token: x, price: res_body["data"][x]["last_price"]} end)
  end

  def ltp(instrument) when is_integer(instrument) do
    ltp(Integer.to_string(instrument))
  end
        
  def ltp(instrument) when is_binary(instrument) do
    url = KiteConnect.gen_url("quote.ltp", instrument)
    headers = KiteConnect.gen_headers()
    res_body = KiteConnect.HTTP.get(url, headers)
    res_body["data"][instrument]["last_price"]
  end

  @doc """
  Full market quote of instruments
  """
  def market_quote(instrument) when is_integer(instrument) do
    market_quote(Integer.to_string(instrument))
  end
        
  def market_quote(instrument) when is_binary(instrument) do
    url = KiteConnect.gen_url("quote.market", instrument)
    headers = KiteConnect.gen_headers()
    res_body = KiteConnect.HTTP.get(url, headers)
  end

  @doc """
  Get bid/ask for instrument
  """
  def get_bid_ask(instrument) when is_integer(instrument) do
    get_bid_ask(Integer.to_string(instrument))
  end
        
  def get_bid_ask(instrument) when is_binary(instrument) do
    body = market_quote(instrument)
    depth = body["data"][instrument]["depth"]

    {
      :bid, Enum.at(depth["buy"], 0)["price"], 
      :ask, Enum.at(depth["sell"], 0)["price"]
    }
  end
end


lib/kite_connect.ex
defmodule KiteConnect do
  @moduledoc """
  Elixir module for the Zerodha Kite Connect API

  ## Installation

  Add `kite_connect` to your list of dependencies in `mix.exs`:
  
  ```elixir
  def deps do
  [
    {:kite_connect, "~> 0.1"}
  ]
  end
  ```

  ## Setup

  Your project should be an OTP application. In your `application.ex`
  file start `KiteConnect.State` as a worker.

  ```elixir
  # Define workers and child supervisors to be supervised
  children = [
  
    # Start your own worker
    worker(KiteConnect.State, []),
  ]

  ```

  A sample iex session:

  ```
  iex(1)> KiteConnect.init(your_api_key, your_api_secret)
  :ok
  iex(2)> KiteConnect.set_access_token(your_request_token)
  :ok
  iex(3)> KiteConnect.Quote.ltp("NSE:INFY")
  724.3
  ```
  """

  @module_map %{
    "quote.ltp" => "quote/ltp",
    "quote.market" => "quote",
    "session.token" => "session/token",
    "user.margins" => "user/margins",
    "instruments.historical" => "instruments/historical",
    "orders.regular" => "orders/regular",
    "orders.bo" => "orders/bo",
    "orders.history" => "orders",
    "orders" => "orders",
  }

  def init(api_key, api_secret) do
    KiteConnect.State.set(:api_key, api_key)
    KiteConnect.State.set(:api_secret, api_secret)
  end
  
  def api_key do
    KiteConnect.State.get(:api_key)
  end
  
  def api_secret do
    KiteConnect.State.get(:api_secret)
  end
  
  def access_token do
    KiteConnect.State.get(:access_token)
  end
  
  def set_access_token(request_token) do
    {:ok, at} = gen_access_token(request_token)
    KiteConnect.State.set(:access_token, at)
  end
  
  def gen_url(module, a1 \\ "", a2 \\ "") do
    api_endpoint = "https://api.kite.trade"

    case module do
      "quote.ltp" ->
        case is_list(a1) do
          true ->
            "#{api_endpoint}/#{@module_map[module]}?i=" <> Enum.join(a1, "&i=")
          false ->
            "#{api_endpoint}/#{@module_map[module]}?i=#{a1}"
        end

      "quote.market" ->
        case is_list(a1) do
          true ->
            "#{api_endpoint}/#{@module_map[module]}?i=" <> Enum.join(a1, "&i=")
          false ->
            "#{api_endpoint}/#{@module_map[module]}?i=#{a1}"
        end

      "instruments.historical" ->
        "#{api_endpoint}/#{@module_map[module]}/#{a1}/#{a2}"

      "orders.history" ->
        "#{api_endpoint}/#{@module_map[module]}/#{a1}"

      _ ->
        "#{api_endpoint}/#{@module_map[module]}"
    end
  end

  def gen_headers do
    [
      "X-Kite-Version": 3,
      Authorization: "token #{KiteConnect.api_key}:#{KiteConnect.access_token}"
    ]
  end

  def gen_headers_post do
    [
     "Content-Type": "application/x-www-form-urlencoded",
      "X-Kite-Version": 3,
      Authorization: "token #{KiteConnect.api_key}:#{KiteConnect.access_token}"
    ]
  end

  defp gen_access_token(request_token) do
    checksum =
      Base.encode16(
        :crypto.hash(
          :sha256,
          "#{KiteConnect.api_key}#{request_token}#{KiteConnect.api_secret}"
        )
      )
      |> String.downcase()

    url = KiteConnect.gen_url("session.token")
    headers = ["X-Kite-Version": 3, "Content-Type": "application/x-www-form-urlencoded"]

    body =
      %{"api_key" => KiteConnect.api_key, "request_token" => request_token, "checksum" => checksum}
      |> URI.encode_query()

    res_body = KiteConnect.HTTP.post(url, body, headers)
    at = res_body["data"]["access_token"]
    
    cond do
      at != nil -> {:ok, at}
      at == nil -> {:error, nil}
    end
  end
end
